### ISSUE TITLE
[ssr]: document `exact <: term`

### ISSUE THREAD

**Issue by gares** [coq/coq#6880]
While cleaning up ssr I noticed that
```coq
exact <: term.
```
is not documented *and* it calls `exact_no_check`.

My take is that we should document it but also make it "safe" (remove the `no_check`).
It seems more in line with the spirit of ssr.

@ggonthier any thoughts?

**Comment by maximedenes** [comment]
When / how was it added, and what were the motivations?

**Comment by gares** [comment]
I think @amahboubi may also be interested in this issue

**Comment by ggonthier** [comment]
I think it goes back to the four color theorem proof. The purpose was to try to optimise the computational proofs, so that the computation used the byte code interpreter, and was not done multiple times (hence the no_check).

**Comment by gares** [comment]
It also serves the purpose of inserting a VMCast without typing the type of the goal.
```
Goal if true is true then True else False.
1 subgoal
  
  ============================
  if true then True else False

Unnamed_thm < exact: (I <: _).
Toplevel input, characters 0-15:
> exact: (I <: _).
> ^^^^^^^^^^^^^^^
Error: Ltac call to "exact (ssrexactarg)" failed.
       Cannot apply lemma (I <: _)
```
The failure is not specific `exact:`, also `exact` fails.

The fact you don't have to repeat the goal is interesting per se (and the reason why I was advertising it on gitter yesterday), but orthogonal to the `no_check` (that I missed completely).

Given that it is not documented I guess we have the luxury to change its behavior, or provide both the safe and the quick variants (but I need a syntax for that!).

**Comment by amahboubi** [comment]
Hum,  without ssr loaded,

`Goal if true then True else False.
  exact (I <: _).`

answers:
`Error: Cannot check cast with vm: unresolved arguments remain.`

can someone explain me why? (`simpl` does not help).

**Comment by gares** [comment]
I guess it does try to call vm before unifying the type of the cast `_` with the goal.

**Comment by ggonthier** [comment]
The idea of `exact <:` was to be a "fast `exact:`"; if you care about safety rather than speed, you can use the latter. I guess the safe option might be useful for debugging; perhaps controlling this via an option might be better? Also, do you think we should support native casts as well?

**Comment by gares** [comment]
I think so, and the syntax is somewhat obvious `exact <<: term`.
For the option thing, I guess I'll ask the users that reported the "issue":
 @Janno yesterday where you using `exact<:` interactive or in an ltac script that has to fail if the goal does not convert to `True`?

**Comment by amahboubi** [comment]
@gares : when I write that `simpl` does not help, I mean that:

```Goal True.
1 subgoal

  ============================
  True
Unnamed_thm < exact (I <: _).
```
also raises
`Error: Cannot check cast with vm: unresolved arguments remain.`

**Comment by gares** [comment]
Yes, I think exact passes to vm_compute `_` directly, before unifying it with the goal, and vm_compute does not like implicit arguments.

**Comment by amahboubi** [comment]
Ok. But then, since my goal is `True`, I miss the point you made that this syntax is used to: "also serves the purpose of inserting a VMCast without typing the type of the goal."  But this is not the topic of the thread anyway.

Back to the documentation issue: if we document the `<:` and `<<:` syntax, then it seems this will be the only place in the manual referring to the `*no_check` tactics, is this ok?

**Comment by gares** [comment]
If you type `exact (I <: if true is true then True else False)` I think it works, this is what I meant. With `exact<: I` you don't need to repeat the goal (and you get the `<:` vm cast)

**Comment by amahboubi** [comment]
Thanks! Now I understand. You meant that the ssr version can be used for this purpose, that's what I misunderstood.

**Comment by Zimmi48** [comment]
@gares Do you intend to address this in the foreseeable future? If not, please remove your assignment.

**Comment by gares** [comment]
removed

**Comment by JasonGross** [comment]
I can't quite tell from the discussion: is `exact <: foo` just a more concise alias for `vm_cast_no_check foo`?  And the suggestion here is to make it a more concise alas for `vm_compute; exact foo`?

**Comment by ppedrot** [comment]
FTR, `exact <: foo` is exactly `exact_no_check (foo <: T)` where `T` is the conclusion. So, it does not check anything but the kernel will use the VM to check the conversion. Said otherwise, that's literally the same thing as `vm_cast_no_check`. Why don't we simply deprecate the former in favour of the latter?

**Comment by SkySkimmer** [comment]
>FTR, exact <: foo is exactly exact_no_check (foo <: T) where T is the conclusion.

That's wrong,
~~~coq
Goal False.
  exact_no_check (I <: False).
~~~
fails but `exact <: I` succeeds.

**Comment by ppedrot** [comment]
It's true, it's just that what I wrote is not in the same context. Namely, you're writing Ltac code where the argument of tactics is typechecked, so this will fail. I was talking about the low-level implementation. In any case, this is essentially an alias of  `vm_cast_no_check`.
